Bluetooth: GATT: Allow to tune writable device appearance permissions

sjanc opened the pull request:
This allows to select if device appearance requires encryption or authentication or none. This makes it similar to what is implemented for device name characteristic.

Thalley commented on subsys/bluetooth/host/Kconfig.gatt:
Arguably the device name, appearance, central address resolution etc. belong in a `Kconfig.gap` file. 

That is outside the scope of this PR, but something we should consider for the future.

For now it makes sense to place the device name GATT options and appearance next to each other

Thalley commented on subsys/bluetooth/host/Kconfig.gatt:
```diff
@@ -284,6 +284,31 @@ config DEVICE_NAME_GATT_WRITABLE_AUTHEN
 
+config BT_DEVICE_APPEARANCE_GATT_WRITABLE
+	bool "Allow to set GAP Appearance by remote GATT clients"
+	depends on BT_DEVICE_APPEARANCE_DYNAMIC
+	default y
+	help
```
```suggestion
	depends on BT_DEVICE_APPEARANCE_DYNAMIC
	help
```

I don't think this should be the default

sjanc replied:
ok, I wanted to keep this consistent with name, but I'm fine with that

Thalley replied:
Yeah, making the GAP GATT attributes similar is something that definitely makes sense. Maybe the name needs to match this instead, but that's outside the scope of this :)

Thalley commented on subsys/bluetooth/host/Kconfig.gatt:
```diff
@@ -284,6 +284,31 @@ config DEVICE_NAME_GATT_WRITABLE_AUTHEN
+	  Enabling this option allows remote GATT clients to write to device
+	  appearance GAP characteristic.
+
+if BT_DEVICE_APPEARANCE_GATT_WRITABLE
+	config BT_DEVICE_APPEARANCE_GATT_WRITABLE_ENCRYPT
+	bool "Encryption required to write appearance by remote GATT clients"
+	default y
+	help
+	  Enabling this option requires the connection to be encrypted to write
+	  to the device appearance GAP characteristic.
+
+
+config DEVICE_APPEARANCE_GATT_WRITABLE_AUTHEN
+	bool "Authentication required to write appearance by remote GATT clients"
+	help
+	  Enabling this option requires the connection to be encrypted and
+	  authenticated to write to the device appearance GAP characteristic.
```
Should `DEVICE_APPEARANCE_GATT_WRITABLE_AUTHEN` depend on `BT_DEVICE_APPEARANCE_GATT_WRITABLE`? Or should these be a `choice`? 

I see that it's following the name above, so I guess it's OK and would otherwise need to be fixed for both. 

Since you are modifying the default of the characteristic (going from authenticated to just encryption), we should mentioned that in the release notes, or default to the same value

sjanc replied:
I kept this consistent with name,  didn't wanted to over-engineer this, also I can keep authenticated as default

alwa-nordic commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -174,9 +174,15 @@ static ssize_t write_appearance(struct bt_conn *conn, const struct bt_gatt_attr
+#elif defined(CONFIG_BT_DEVICE_APPEARANCE_GATT_WRITABLE_ENCRYPT)
+	#define GAP_APPEARANCE_PERMS (BT_GATT_PERM_READ | BT_GATT_PERM_WRITE_ENCRYPT)
+#else
+	#define GAP_APPEARANCE_PERMS (BT_GATT_PERM_READ | BT_GATT_PERM_WRITE)
```
It's a bit dangerous that the default is the most permissive? It is also not backwards-compatible. The default should be `BT_GATT_PERM_WRITE_AUTHEN`, like it was before.

How about we define a kconfig choice defaulting to 'authenticated'?

sjanc committed:
Bluetooth: GATT: Convert writable device name to KConfig choice

This makes it clear on how security requirements are set and avoid
confusion on how encryption and authentication are handled.

Signed-off-by: Szymon Janc <szymon.janc@codecoup.pl>

sjanc committed:
Bluetooth: GATT: Allow to tune writable device appearance permissions

This allows to select if device appearance requires encryption or
authentication or none. This makes it similar to what is implemented
for device name characteristic.

Signed-off-by: Szymon Janc <szymon.janc@codecoup.pl>

sjanc commented:
converted to choice and defaults are set to what was already in tree (name enc, appearance enc and auth)
